File: app/src/main/java/com/bis5/fitjourney/viewmodels/WorkoutViewModel.java
package com.bis5.fitjourney.viewmodels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;

import com.bis5.fitjourney.models.AppDatabase;
import com.bis5.fitjourney.models.Exercise;
import com.bis5.fitjourney.models.ExerciseDao;
import com.bis5.fitjourney.models.SetLog;
import com.bis5.fitjourney.models.SetLogDao;
import com.bis5.fitjourney.models.Workout;
import com.bis5.fitjourney.models.WorkoutDao;
import com.bis5.fitjourney.models.WorkoutLog;
import com.bis5.fitjourney.models.WorkoutLogDao;
import com.bis5.fitjourney.other.WorkoutWithExercises;

import java.util.List;
import java.util.UUID;

public class WorkoutViewModel extends ViewModel {

    private final WorkoutDao workoutDao;
    private final ExerciseDao exerciseDao;
    private final WorkoutLogDao workoutLogDao;
    private final SetLogDao setLogDao;

    public final MutableLiveData<String> userEmailTrigger = new MutableLiveData<>();

    public final LiveData<WorkoutWithExercises> mostRecentWorkout;
    public final LiveData<List<Workout>> userWorkouts;

    public WorkoutViewModel(WorkoutDao workoutDao, ExerciseDao exerciseDao, WorkoutLogDao workoutLogDao, SetLogDao setLogDao) {
        this.workoutDao = workoutDao;
        this.exerciseDao = exerciseDao;
        this.workoutLogDao = workoutLogDao;
        this.setLogDao = setLogDao;

        mostRecentWorkout = Transformations.switchMap(userEmailTrigger, workoutDao::getMostRecentWorkoutWithExercises);
        userWorkouts = Transformations.switchMap(userEmailTrigger, workoutDao::getWorkoutsForUser);
    }

    public void loadUser(String userEmail) {
        if (userEmail != null && !userEmail.equals(userEmailTrigger.getValue())) {
            userEmailTrigger.setValue(userEmail);
        }
    }

    public LiveData<Workout> getWorkout(long workoutId) {
        return workoutDao.getWorkout(workoutId);
    }

    public LiveData<List<Exercise>> getExercises(long workoutId) {
        return exerciseDao.getExercisesForWorkout(workoutId);
    }

    public LiveData<List<SetLog>> getLoggedSets(String workoutLogId) {
        return setLogDao.getLoggedSetsForWorkout(workoutLogId);
    }

    public void createWorkout(String name, String userEmail) {
        AppDatabase.databaseWriteExecutor.execute(() -> {
            Workout newWorkout = new Workout(name, name, System.currentTimeMillis(), 0, userEmail);
            workoutDao.insert(newWorkout);
        });
    }

    public void updateWorkoutName(long workoutId, String newName) {
        AppDatabase.databaseWriteExecutor.execute(() -> workoutDao.updateWorkoutName(workoutId, newName));
    }

    public void deleteWorkout(Workout workout) {
        AppDatabase.databaseWriteExecutor.execute(() -> workoutDao.delete(workout));
    }

    public void addExercise(long workoutId, String name, int sets, int reps) {
        AppDatabase.databaseWriteExecutor.execute(() -> {
            Exercise newExercise = new Exercise(workoutId, name, sets, reps);
            exerciseDao.insert(newExercise);
        });
    }

    public LiveData<String> startWorkout(long workoutId) {
        MutableLiveData<String> workoutLogIdLiveData = new MutableLiveData<>();
        AppDatabase.databaseWriteExecutor.execute(() -> {
            String userEmail = workoutDao.getUserEmailForWorkout(workoutId);
            if (userEmail == null) userEmail = "";

            WorkoutLog workoutLog = new WorkoutLog(UUID.randomUUID().toString(), workoutId, userEmail, System.currentTimeMillis(), null, "in-progress");
            workoutLogDao.insert(workoutLog);
            workoutLogIdLiveData.postValue(workoutLog.getId());
        });
        return workoutLogIdLiveData;
    }

    public LiveData<List<WorkoutLog>> getWorkoutHistory(long workoutId) {
        return workoutLogDao.getLogsForWorkout(workoutId);
    }

    public LiveData<List<WorkoutLog>> getAllLogsForUser(String userEmail) {
        return workoutLogDao.getAllLogsForUser(userEmail);
    }

    public LiveData<List<SetLog>> getSetHistoryForExercise(String userEmail, String exerciseName) {
        return setLogDao.getSetHistoryForExercise(userEmail, exerciseName);
    }

    public LiveData<List<String>> getUniqueExerciseNames(String userEmail) {
        return setLogDao.getUniqueLoggedExerciseNames(userEmail);
    }

    public void logSet(String workoutLogId, String exerciseName, int setNumber, int reps, double weight) {
        AppDatabase.databaseWriteExecutor.execute(() -> {
            SetLog setLog = new SetLog(UUID.randomUUID().toString(), workoutLogId, exerciseName, setNumber, reps, weight, System.currentTimeMillis(), "completed");
            setLogDao.insert(setLog);
        });
    }

    public void finishWorkout(String logId) {
        AppDatabase.databaseWriteExecutor.execute(() -> workoutLogDao.finishWorkout(logId, System.currentTimeMillis()));
    }
}

File: app/src/main/java/com/bis5/fitjourney/viewmodels/WorkoutViewModel.kt
package com.bis5.fitjourney.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.asLiveData
import androidx.lifecycle.liveData
import androidx.lifecycle.viewModelScope
import com.bis5.fitjourney.models.Exercise
import com.bis5.fitjourney.models.ExerciseDao
import com.bis5.fitjourney.models.SetLog
import com.bis5.fitjourney.models.SetLogDao
import com.bis5.fitjourney.models.Workout
import com.bis5.fitjourney.models.WorkoutDao
import com.bis5.fitjourney.models.WorkoutLog
import com.bis5.fitjourney.models.WorkoutLogDao
import kotlinx.coroutines.launch

class WorkoutViewModel(private val workoutDao: WorkoutDao, private val exerciseDao: ExerciseDao, private val workoutLogDao: WorkoutLogDao, private val setLogDao: SetLogDao) : ViewModel() {

    fun getWorkouts(userEmail: String): LiveData<List<Workout>> {
        return workoutDao.getWorkoutsForUser(userEmail)
    }

    fun getWorkout(workoutId: String): LiveData<Workout?> {
        return workoutDao.getWorkout(workoutId)
    }

    fun getExercises(workoutId: String): LiveData<List<Exercise>> {
        return exerciseDao.getExercisesForWorkout(workoutId).asLiveData()
    }

    fun getLoggedSets(workoutLogId: String): LiveData<List<SetLog>> {
        return setLogDao.getLoggedSetsForWorkout(workoutLogId)
    }

    fun createDefaultWorkoutsIfNoneExist(userEmail: String) {
        viewModelScope.launch {
            val workoutCount = workoutDao.getWorkoutCountForUser(userEmail)
            if (workoutCount == 0) {
                val defaultWorkouts = listOf(
                    Workout(name = "Push Day", userEmail = userEmail, date = System.currentTimeMillis(), workoutType = "Push Day", duration = 0),
                    Workout(name = "Pull Day", userEmail = userEmail, date = System.currentTimeMillis(), workoutType = "Pull Day", duration = 0),
                    Workout(name = "Leg Day", userEmail = userEmail, date = System.currentTimeMillis(), workoutType = "Leg Day", duration = 0),
                    Workout(name = "Upper Body", userEmail = userEmail, date = System.currentTimeMillis(), workoutType = "Upper Body", duration = 0),
                    Workout(name = "Lower Body", userEmail = userEmail, date = System.currentTimeMillis(), workoutType = "Lower Body", duration = 0),
                    Workout(name = "Core", userEmail = userEmail, date = System.currentTimeMillis(), workoutType = "Core", duration = 0),
                    Workout(name = "Cardio", userEmail = userEmail, date = System.currentTimeMillis(), workoutType = "Cardio", duration = 0)
                )
                defaultWorkouts.forEach { workoutDao.insertWorkout(it) }
            }
        }
    }

    fun createWorkout(name: String, userEmail: String) {
        viewModelScope.launch {
            val newWorkout = Workout(
                name = name, 
                userEmail = userEmail, 
                date = System.currentTimeMillis(),
                workoutType = name,
                duration = 0
            )
            workoutDao.insertWorkout(newWorkout)
        }
    }

    fun updateWorkoutName(workoutId: String, newName: String) {
        viewModelScope.launch {
            workoutDao.updateWorkoutName(workoutId, newName)
        }
    }

    fun deleteWorkout(workout: Workout) {
        viewModelScope.launch {
            workoutDao.deleteWorkout(workout)
        }
    }

    fun addExercise(workoutId: String, name: String, sets: Int, reps: Int, weight: Double) {
        viewModelScope.launch {
            val newExercise = Exercise(workoutId = workoutId, name = name, sets = sets, reps = reps, weight = weight)
            exerciseDao.insertExercise(newExercise)
        }
    }

    fun startWorkout(workoutId: String): LiveData<String> = liveData {
        val userEmail = workoutDao.getUserEmailForWorkout(workoutId) ?: ""
        val workoutLog = WorkoutLog(workoutId = workoutId, userEmail = userEmail, dateStarted = System.currentTimeMillis())
        val workoutLogId = workoutLogDao.startWorkout(workoutLog)
        emit(workoutLogId.toString())
    }

    fun getWorkoutHistory(workoutId: String): LiveData<List<WorkoutLog>> {
        return workoutLogDao.getLogsForWorkout(workoutId)
    }

    fun getAllLogsForUser(userEmail: String): LiveData<List<WorkoutLog>> {
        return workoutLogDao.getAllLogsForUser(userEmail)
    }

    fun getSetHistoryForExercise(userEmail: String, exerciseName: String): LiveData<List<SetLog>> {
        return setLogDao.getSetHistoryForExercise(userEmail, exerciseName)
    }

    fun getUniqueExerciseNames(userEmail: String): LiveData<List<String>> {
        return setLogDao.getUniqueLoggedExerciseNames(userEmail)
    }

    fun logSet(workoutLogId: String, exerciseName: String, setNumber: Int, reps: Int, weight: Double) {
        viewModelScope.launch {
            val setLog = SetLog(
                workoutLogId = workoutLogId,
                exerciseName = exerciseName,
                setNumber = setNumber,
                reps = reps,
                weight = weight,
                status = "completed"
            )
            setLogDao.logSet(setLog)
        }
    }

    fun finishWorkout(logId: String) {
        viewModelScope.launch {
            workoutLogDao.finishWorkout(logId, System.currentTimeMillis())
        }
    }
}
